### ISSUE TITLE
Chords / layers

### ISSUE THREAD

**Issue by marcos-diaz** [inputlabs/alpakka_firmware#70]
Chords: Being able to set combination of buttons (eg `L1+R1`) with their own binds.

Layers: Being able to set a "shift" button that changes the behaviour of several other buttons, eg while holding `L1` then `A, B, X, Y` have different binds.

Since the Alpakka firmware runs on embedded hardware, there will be some limitations compared with Steam Input, for example we would have a limited number of chords and layers (around 4) per profile.

**Comment by Electronicks** [comment]
I am willing to help coding this and possibly more features. I wrote JoyShockMapper's button handling, and it is one of, if not THE most comprehensive and flexible remapper.

My current implementation in JSM is a state machine getting button state as input (as well as timestamp, settings, etc...) and key up/down events as output. My setup includes chorded press, simultaneous press, "diagonal" press, double press, instant/pulse press, toggle presses, turbo press. All of the presses. :D

I understand coding for the raspi is done in good ol' C, so my C++ would need some adaptation, in addition to memory usage concern.

My own Alpakka 0.97 is not built yet : the PCB is on a boat and who knows where my digikey order is. I'm eager to play with it hopefully by the end of the month.

I'm on the discord too, if you want to PM me. Cheers

**Comment by Electronicks** [comment]
To follow up from the Discord conversation, implementing this is a sizeable amount of change and it is better to break it down in small parts you can review and approve. Is this the best place to work together on this? Or would you rather Discord DM, or Discord Thread?

The first thing I would like to pass by you is the changes to the action mapping. Currently you have a button mode identifier and three arrays of buttons that are triggered at different times depending on the mode.

The setup I would like to create instead is what I have in JSM: instead of calling the hid function directly from the button, the button instead would raise events and lookup in the mapping for actions/callbacks that are mapped for that event. A Mapping thus binds button events to one or more actions/callbacks that are constructed from 3 parts: an action modifier (press, pulse, toggle, release), the key press and an event modifier (on press, tap, hold, turbo, release).

For example, one mapping array could be "Toggle 'RMOUSE' on tap and Press 'RMOUSE' on hold.

Then, a whole separate set of mappings is used for double presses, enabling double press taps and double press holds.

This would mean that profiles needs to store a collection of those three parts, at least twice: the first for the basic press and the second for double press.

Once we have this in place it will be fairly easy to add more mappings for chords, layers and more kinds of presses. But I am slightly worried on memory usage if we use predetermined tables for each of those. That's why I was hoping to use a linked list to have dynamic collection sizes. But I'll start with predetermined tables first and see where this goes.

What do you think?

**Comment by rempas** [comment]
Please, also add a way to "toggle" layouts in addition to *hold* the key down. For example, let's say that *normally*:

`A` generates the key **J**
`B` generates the key **k**
`X` generates the key **h**
`Y` generates the key **l**

Now, allow for, let's say `L1+A` to toggle the second "present" (better not call the "togglable" ones layers as well) where:

`A` generates the key **a**
`B` generates the key **b**
`X` generates the key **x**
`Y` generates the key **y**

To extend this, those "presets" could even change behaviors of the controller. For example, in one of them, the right stick acts like a mouse and the other, it acts like flickstick. Or, where the mouse wheel is used to cycle between weapons/items or to turn your camera. Now, that's not very necessary, but it would be great.

In conclusion, we'll have a combination (like `L1+A`) to *toggle* presets and a single button to *hold* layers. Similar to how Steam works, if I remember correctly. Toggling will allow us to be able to reach every button comfortable and use the controller as a fully capable keyboard while *holding* the layer will help us do some additional *fast* changes (maybe the example with the right stick makes more sense in layers).

Maybe what I propose is what you mean with *chords*? I don't understand what that description means exactly. But I think it's a good idea to write this and talk about it and you let me know.

**Comment by Electronicks** [comment]
Aren't "toggleable layers" something you can replicate with a separate profile already?

- There already is the ability to copy profile from one to another
- Perform "layer changes" on the new profile
- In the original profile, assign binding to change to the new profile
- In the new profile assign button to change to the original profile

**Comment by rempas** [comment]
> Aren't "toggleable layers" something you can replicate with a separate profile already?
> 
> * There already is the ability to copy profile from one to another
> * Perform "layer changes" on the new profile
> * In the original profile, assign binding to change to the new profile
> * In the new profile assign button to change to the original profile

I don't own an Alpakka controller so, I don't know exactly how profiles work. If they work exactly as you said, then sure! I don't see any limitation or problem with that!

**Comment by Jazztache** [comment]
Steam-Input style mode shifting isn't possible in Ctrl as of now to my knowledge. I tried making separate profiles, and whilst toggling them on did work, there is no way to hold a button down to hold down a profile. This would be very nice to have, and yes, this is something that is in JSM.

You could make it so that releasing a button is binding as well, which can then trigger a switch back to a profile, but that seems a little bit clunky and not the cleanest way to implement it.

**Comment by Electronicks** [comment]
For buttons I would probably go the way I do in JSM, which is an alternate mapping being activate when the chord is active. Then I just have to keep track of the order buttons are pressed and held, which is easy in C++ but much harder in C.
